### ISSUE TITLE
Enable offsets to map position within viewport

### ISSUE THREAD

**Issue by matteblair** [tangrams/tangram-es#841]
This is a common need in navigation applications. When displaying a navigation UI, it is often useful to track the user's current location and place the user's current location in the lower portion of the display in order to show more area in the direction of travel. 

Currently in tangram-es there is no simple mechanism to do this - the location given to `setPosition()` will always be in the center of the display. The only work-around is to call `setPosition()` with a location that is ahead of the user's location in the direction of travel. 

A possible interface for this could be:

``` c++
// Set the displacement of the map's position from the center of the display.
// The location set as the map position will appear at the screen position 
// (w/2 + xPixels, h/2 + yPixels) where 'w' and 'h' are the viewport width and height.
void setViewOffset(int xPixels, int yPixels);
```

**Comment by tallytalwar** [comment]
I think the equivalent google maps (javascript) api is:

```
panBy(x:number, y:number)	
Return Value:  None
Changes the center of the map by the given distance in pixels. If the distance is less 
than both the width and height of the map, the transition will be smoothly animated. 
Note that the map coordinate system increases from west to east (for x values) and 
north to south (for y values).
```
Reference: https://developers.google.com/maps/documentation/javascript/reference

**Comment by tallytalwar** [comment]
Nopes, thats not the equivalent api call!

**Comment by westnordost** [comment]
This would be very handy. Currently, I have to do this myself and the code is quite messy. I'd suggest however a method like this:
```java
void setViewOffsets(int left, int top, int right, int bottom)
```

My use case is that I want the view to zoom to fully show a geometry, which could also be a large building or a road. Since I must calculate the zoom that still fits the geometry, I need to know how large the view window (with offsets is).

This is my code (shortened a bit):
```java
	private void zoomAndMoveToContain(ElementGeometry g)
	{
		float targetZoom = getMaxZoomThatContains(g);
		float currentZoom = controller.getZoom();

		controller.setZoom(targetZoom);
		LngLat pos = getCenterWithOffset(g);
		controller.setZoom(currentZoom);

		if(pos != null) controller.setPositionEased(pos, 500);
		controller.setZoomEased(targetZoom, 500);
	}

	private float getMaxZoomThatContains(ElementGeometry geometry)
	{
		BoundingBox objectBounds = geometry.getBounds();
		BoundingBox screenArea;
		float currentZoom;
		synchronized(controller) {
			screenArea = getDisplayedArea(viewOffsets);
			if(screenArea == null) return Float.NaN;
			currentZoom = controller.getZoom();
		}

		double screenWidth = screenArea.getMaxLongitude() - screenArea.getMinLongitude();
		double screenHeight = screenArea.getMaxLatitude() - screenArea.getMinLatitude();

		double objectWidth = objectBounds.getMaxLongitude() - objectBounds.getMinLongitude();
		double objectHeight = objectBounds.getMaxLatitude() - objectBounds.getMinLatitude();

		double zoomDeltaX = Math.log10(screenWidth / objectWidth) / Math.log10(2.);
		double zoomDeltaY = Math.log10(screenHeight / objectHeight) / Math.log10(2.);

		return (float) Math.max(1, currentZoom + Math.min(zoomDeltaX, zoomDeltaY));
	}

	public BoundingBox getDisplayedArea(Rect offset)
	{
		if(controller == null) return null;
		if(getView() == null) return null;
		Point size = new Point(
				getView().getWidth() - offset.left - offset.right,
				getView().getHeight() - offset.top - offset.bottom);
		if(size.equals(0,0)) return null;

		// the special cases here are: map tilt and map rotation:
		// * map tilt makes the screen area -> world map area into a trapezoid
		// * map rotation makes the screen area -> world map area into a rotated rectangle

		// dealing with tilt: this method is just not defined if the tilt is above a certain limit
		if(controller.getTilt() > Math.PI / 4f) return null; // 45°

		LatLon[] positions = new LatLon[4];
		positions[0] = controller.screenPositionToLngLat(new PointF(offset.left,          offset.top));
		positions[1] = controller.screenPositionToLngLat(new PointF(offset.left + size.x ,offset.top));
		positions[2] = controller.screenPositionToLngLat(new PointF(offset.left,          offset.top + size.y));
		positions[3] = controller.screenPositionToLngLat(new PointF(offset.left + size.x, offset.top + size.y));

		// dealing with rotation: find each the largest latlon and the smallest latlon, that'll
		// be our bounding box

		Double latMin = null, lonMin = null, latMax = null, lonMax = null;
		for (LatLon position : positions)
		{
			if(position == null) return null;
			double lat = position.getLatitude();
			double lon = position.getLongitude();

			if (latMin == null || latMin > lat) latMin = lat;
			if (latMax == null || latMax < lat) latMax = lat;
			if (lonMin == null || lonMin > lon) lonMin = lon;
			if (lonMax == null || lonMax < lon) lonMax = lon;
		}

		return new BoundingBox(latMin, lonMin, latMax, lonMax);
	}

	private LngLat getCenterWithOffset(ElementGeometry geometry)
	{
		int w = getView().getWidth();
		int h = getView().getHeight();

		LngLat normalCenter = controller.screenPositionToLngLat(new PointF(w/2f, h/2f));

		LngLat offsetCenter = controller.screenPositionToLngLat(new PointF(
				viewOffsets.left + (w - viewOffsets.left - viewOffsets.right)/2,
				viewOffsets.top + (h - viewOffsets.top - viewOffsets.bottom)/2));

		if(normalCenter == null || offsetCenter == null) return null;

		LngLat pos = TangramConst.toLngLat(geometry.center);
		pos.latitude -= offsetCenter.latitude - normalCenter.latitude;
		pos.longitude -= offsetCenter.longitude - normalCenter.longitude;
		return pos;
	}

	public void setViewOffsets(Rect offsets)
	{
		viewOffsets = offsets;
	}
```

**Comment by matteblair** [comment]
@westnordost Does this satisfy your use case?
```java
CameraUpdateFactory.newLatLngBounds(LngLat sw, LngLat ne, Rect padding)
```
([code](https://github.com/tangrams/tangram-es/blob/master/platforms/android/tangram/src/main/java/com/mapzen/tangram/CameraUpdateFactory.java#L165))

It seems like you already get the bounding box for your subject, and set your view offsets as a Rect.

**Comment by westnordost** [comment]
It does, fully, and with a better interface than I could have hoped for. :-o

**Comment by matteblair** [comment]
I think `CameraUpdateFactory.newLatLngBounds` can fill this need for many uses, but not all. Another common need is to focus the view on a point with padding. `newLngLatBounds` can't do this because a point is an infinitely small bounds, so calling this method with the same point as both corners of the bounds will always result in a camera at the maximum zoom. Padding should be accounted for in other camera behaviors too, e.g. when zooming a padded view the focus point of the zoom should be the center of the padded area.

In short, I think we still need to implement something like `setViewPadding(Rect padding)`.

**Comment by nvkelso** [comment]
Uber's recent [blog post](https://eng.uber.com/building-a-scalable-and-reliable-map-interface-for-drivers/) highlights there addition of a `Camera Director` to keep track of all the different (and sometimes contradictory) demands on a view.

**Comment by whitespeed** [comment]
I cannot find setViewPadding(Rect padding), does it exist?

**Comment by matteblair** [comment]
Ah, no it hasn't been implemented yet.

**Comment by liodali** [comment]
is there any way to get boundingbox or we should calculated from center position in iOS ?

**Comment by matteblair** [comment]
@liodali You can determine the view bounds by calling `coordinateFromViewPosition` on the four corners of your `TGMapView`. There isn't a convenience method to do this for you at the moment, but that might be added in the future.

**Comment by matteblair** [comment]
Also, the original issue here has been resolved by the addition of the `setPadding` functions (https://github.com/tangrams/tangram-es/pull/2294). Closing this issue!
